runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunner.java

```diff
@@ -120,7 +121,9 @@ public PipelineResult run(Pipeline pipeline) {
 
     if (result instanceof DetachedEnvironment.DetachedJobExecutionResult) {
       LOG.info("Pipeline submitted in Detached mode");
-      return new FlinkDetachedRunnerResult();
+      FlinkDetachedRunnerResult flinkDetachedRunnerResult = new FlinkDetachedRunnerResult();
+      // no metricsPusher because metrics are not supported in detached mode
```

Comment:
@echauchot Does this mean that any metrics created with `Metrics.counter(), Metrics.gauge()` etc. methods in classes extending `DoFn` won't be visible/computed if the pipeline executes in detached mode?

Reply:
I added this comment because of https://github.com/apache/beam/blob/c0d902f9ba1307571c5b97fdf56765bb70e3e123/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkDetachedRunnerResult.java#L41 it means that metrics could not be requested for the PipelineResult in detached mode because it is not available in detached mode. 
The thing is that MetricsPusher does not use PipelineResult to extract metrics but it uses it to know pipeline execution status to kill the extraction thread. So basically, if I set up a MetricsPusher in detached mode I would not be able to stop the thread leading to leaking thread.
As the whole Flink runner advertise not to support metrics in detached mode, seems coherent to not support metrics extraction either.

Reply:
Flink does support metrics. I guess this exception is thrown only because, before MetricsPusher, the only way to get the metrics out of Beam was to query them from the PipelineResult which is not available in detached mode. That said we could maybe make MetricsPusher work because it uses MetricsContainerStepMap which is available. The only thing is to figure out how to stop the Metrics pusher thread. It might require to completely rewrite MetricsPusher to avoid a pushing thread but rather instrument the pipeline by adding technical DoFns to the DAG that push metrics. But I'm not sure it would be a good thing to "pollute" the DAG that way. Let me check with the Flink guys how we could solve it in detached mode with simpler solution.

Reply:
@echauchot sounds good, please let me know.

Reply:
@zorro786 in Flink there is no driver like there is in Spark. So user main() code (pipeline.run and thud MetricsPusher thread) will run inside the Flink client. In default submission mode, the client lives until the end of pipeline so no issue. In detached mode, the client lives only the time of submission of the job so, as the Metrics thread is a daemon, it would be leaking after client is stopped. If we set it not to be a daemon, then it will stop with the client and no metrics will be pushed. So the current architecture of the MetricsPusher will not support detached mode in Flink. One of the solutions is to rewrite it all using the DAG modification but IMHO I find it overkill to change it all just to support one sub-case on one only runner and more over only metrics pushing and not pulling will be supported. Is there a strong need for supporting metrics export in detached mode even if regular metrics pulling (PipelineResult.metrics()) are not supported?

Reply:
@echauchot There was never support for pulling metrics in detached mode. So when I read about the Metrics Pusher design I thought this will help in overcoming the limitations by providing an alternative way that is runner agnostic. Especially this point "....Even if users will target a particular runner, it would be good for them and for the Beam project to have a coherent way of having the metrics flow no matter the chosen runner....." from your design docs. I would say there is a need for supporting metrics in detached mode (which comes very useful in Flink's standalone deployment mode), whether pushing or pulling needs to be implemented first I don't know.

Reply:
@kennknowles Do you have any comment, I'm hesitating about dropping the current MetricsPusher to allow detached mode, do you see any other possible alternative? Also WDYT about the DAG modification I'm discussing above?

Reply:
I was thinking maybe we can leave it that way and leave the thread in daemon mode. The only thing that is needed is a way to stop the thread without relying on the PipelineResult that is not available in detached mode. In Flink we don't have a callback of the end of a batch pipeline (see https://issues.apache.org/jira/browse/BEAM-3472) but I can maybe plug a thread stop method call in some precise places in the runner.

Reply:
@zorro786 I checked and I'm afraid but supporting metrics pusher in detached mode does not seem to be possible without a modification of the upstream flink. Indeed, in this mode even flink accumulators are not available so we will have no access to the metrics at all, it is not just a matter of terminating the pusher thread as I first thought. To help you understand the architecture, here is a diagram (that is part of my upcoming presentation at the apacheCon)
![metrics inside the runner](https://user-images.githubusercontent.com/8821084/42324124-827864e2-8062-11e8-9396-390f7778ad7d.png)

Reply:
@echauchot Thank you so much for taking time to look in to this. That diagram makes it very clear!

Reply:
@zorro786 I updated the diagram to add the MetricsPusher so that you can have the whole picture. Enjoy !

Reply:
So, do I read this right that detached mode is not going to be supported? Or there is a push to future to make it available there?
Without it, it is not really agnostic way, as there are run modes which does not have metrics. Initially I hoped the integration will follow the same behaviour like if I would implement custom  `org.apache.flink.metrics.reporter.MetricReported`. I do submit my jobs (streaming) in detached mode. I consider it not helpful to introduce client process which can die and what then happens to the application on the cluster and metric pusher.